pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java

```diff
@@ -758,7 +759,23 @@ public void replaceSegment(ImmutableSegment segment, @Nullable ThreadSafeMutable
     }
   }
 
-  protected abstract void removeSegment(IndexSegment segment, MutableRoaringBitmap validDocIds);
+  protected void removeSegment(IndexSegment segment, MutableRoaringBitmap validDocIds) {
+    removeSegment(segment, doGetPrimaryKeyIteratorForRemoveSegment(segment, validDocIds));
+  }
+
+  protected Iterator<PrimaryKey> doGetPrimaryKeyIteratorForRemoveSegment(IndexSegment segment,
+      MutableRoaringBitmap validDocIds) {
+    try (UpsertUtils.PrimaryKeyReader primaryKeyReader = new UpsertUtils.PrimaryKeyReader(segment,
+        _primaryKeyColumns)) {
+      return UpsertUtils.getPrimaryKeyIterator(primaryKeyReader, validDocIds);
```

Comment:
Does this work? Do we need to finish removing all valid docs then close the primary key reader?

Reply:
Great catch! Refactored accordingly.